src/controllers/hooks.js

```diff
@@ -0,0 +1,235 @@
+import { composeBaseURL, hasText } from '@adobe/spacecat-shared-utils';
+
+import { BaseSlackClient, SLACK_TARGETS } from '@adobe/spacecat-shared-slack-client';
+import { SITE_CANDIDATE_STATUS, SITE_CANDIDATE_SOURCES } from '@adobe/spacecat-shared-data-access/src/models/site-candidate.js';
+import { fetch } from '../support/utils.js';
+
+const CDN_HOOK_SECRET_NAME = 'INCOMING_WEBHOOK_SECRET_CDN';
+const RUM_HOOK_SECRET_NAME = 'INCOMING_WEBHOOK_SECRET_RUM';
+
+export const BUTTON_LABELS = {
+  APPROVE_CUSTOMER: 'As Customer',
+  APPROVE_FRIENDS_FAMILY: 'As Friends/Family',
+  IGNORE: 'Ignore',
+};
+
+const IGNORED_SUBDOMAIN_TOKENS = ['demo', 'dev', 'stag', 'qa', '--'];
+
+class InvalidSiteCandidate extends Error {}
+
+function hookAuth(fn, opts) {
+  return (context) => {
+    const expectedSecret = context.env[opts.secretName];
+    const secretFromPath = context.params?.hookSecret;
+    return hasText(expectedSecret) && expectedSecret === secretFromPath
+      ? fn(context)
+      : notFound();
+  };
+}
+
+function errorHandler(fn, opts) {
+  const { type } = opts;
+  return async (context) => {
+    const { log } = context;
+    try {
+      return await fn(context);
+    } catch (e) {
+      if (e instanceof InvalidSiteCandidate) {
+        log.warn(`Could not process the ${type} site candidate. Reason: ${e.message}`);
+        return ok(`${type} site candidate disregarded`);
+      }
+      log.error(`Unexpected error while processing the ${type} site candidate`, e);
+      return internalServerError();
+    }
+  };
+}
+
+function isInvalidSubdomain(hostname) {
+  const subdomain = hostname.split('.').slice(0, -2).join('.');
+  return IGNORED_SUBDOMAIN_TOKENS.some((ignored) => subdomain.includes(ignored));
+}
+
+function isIPAddress(hostname) {
+  return /^\d{1,3}(\.\d{1,3}){3}$|^([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}$/.test(hostname);
+}
+
+function containsPathOrSearchParams(url) {
+  return url.pathname !== '/' || url.search !== '';
+}
+
+async function verifyHelixSite(url) {
+  let finalUrl;
+  try {
+    const resp = await fetch(url);
+    finalUrl = resp.url;
+  } catch (e) {
+    throw new InvalidSiteCandidate(`URL is unreachable ${url}`, { cause: e });
+  }
+
+  finalUrl = finalUrl.endsWith('/') ? `${finalUrl}index.plain.html` : `${finalUrl}.plain.html`;
+  let finalResp;
+
+  try {
+    // redirects are disabled because .plain.html should return 200
+    finalResp = await fetch(finalUrl, { redirect: 'manual' });
+  } catch (e) {
+    throw new InvalidSiteCandidate(`.plain.html is unreachable for ${finalUrl}`, { cause: e });
+  }
+
+  // reject if .plain.html does not return 2XX
+  if (!finalResp.ok) {
+    throw new InvalidSiteCandidate(`.plain.html does not return 2XX for ${finalUrl}`);
+  }
+
+  const respText = await finalResp.text();
+
+  // reject if .plain.html contains <head>
+  if (respText.includes('<head>')) {
+    throw new InvalidSiteCandidate('.plain.html should not contain <head>');
+  }
```

Comment:
could be shortened to:
```suggestion
   try {
     const finalResp = await fetch(finalUrl, { redirect: 'manual' });
     if (!finalResp.ok) {
       throw new InvalidSiteCandidate(`.plain.html does not return 2XX for ${finalUrl}`);
     }
     const respText = await finalResp.text();
     if (respText.includes('<head>')) {
       throw new InvalidSiteCandidate('.plain.html should not contain <head>');
     }
   } catch (e) {
     throw new InvalidSiteCandidate(`.plain.html is unreachable for ${finalUrl}`, { cause: e });
   }
```

Reply:
moving the extra checks inside try blocks would result in inaccurate error messages such as:

```
.plain.html is unreachable for some-url.com
    .plain.html should not contain <head>
```

this log would be wrong when the some-url.com is reachable but also contains a `<head>`. I will keep the `!finalResp.ok` and `respText.includes('<head>')` checks out of the `try` block